src/consumers/consumers.h

```diff
@@ -1,6 +1,84 @@
 #ifndef CONSUMERS_
 #define CONSUMERS_
 
+typedef struct metric_metadata {
+    tag_collection* tags;
+    char* instance;
+    char* sampling;
+} metric_metadata;
+
+typedef struct counter_metric {
+    char* name;
+    metric_metadata* meta;
+    unsigned long long int value;
```

Comment:
I'd probably go for signed 64bit integer, counters can be either increasing or decreasing. Definitely not unsigned that is not safe. @natoscott is 64bits enough for PCP?

Reply:
Are you sure they can be decresing as well? From here https://github.com/statsd/statsd/blob/master/docs/metric_types.md, it seems to be only increasing.

Reply:
In PCP they can be decreasing and it's obvious that anybody can start sending `-1` value, why not to support it? That document is not a standard or anything, it's very likely that the original author forgot about this possibility. It does not make sense to do unsigned, it's such a high number (64bits) that it will be enough for everyone (TM).

Reply:
One constraint is that it must be monotonic, so either increasing or decreasing, but do not check for this, PCP would error out for us.

Reply:
How would that work? I mean if we had an X = 2, and then more updates would arrive such as:
-2
-5
10
Should then result in 5? How is that different, apart type, from gauge then?

Reply:
Question for @natoscott - statsd does not define monotonic counters, it's just free form, however I remember that PCP requires them to be monotonic. How the agent should handle this? Fatal error is one extreme, the other is simply ignoring such a value.

Reply:
> I'd probably go for signed 64bit integer, counters can be either increasing or decreasing. Definitely not unsigned that is not safe. @natoscott is 64bits enough for PCP?

Yes, 64 bits is fine.  FWIW, in C its signed counters that are not safe - behaviour is undefined when wrapping.

Reply:
> Question for @natoscott - statsd does not define monotonic counters, it's just free form, however I remember that PCP requires them to be monotonic. How the agent should handle this? Fatal error is one extreme, the other is simply ignoring such a value.

IIRC statsd counters are reset to zero after each sample is taken? Its probably best if the agent accumulates (sums) all counters then to return them to their natural form, which is what client tools will expect.

Reply:
What I get from this:

- counters need to support both negative and positive values, hence they need to be signed type
- as such they need to have their boundaries checked on each update (this I should implement for gauges as well as they are signed already)
- they need to be monotic as well, but I dont have to check for that as PCP will itself validate data
- 64bit length is fine

So the behaviour should be as such:
-> 0
example:2|c -> 2
example:-5|c -> -3
example:6|c -> 3

I am not sure tho if this is really needed as gauges fill that exact role and it seems to me like feature duplicity.

Gauges example with same results:
-> 0
example:+2|g -> 2
example:-5|g -> -3
example:+6|g -> 3

In my opinion it would be best to leave counters as is and just add signed boundary check to gauges which I should do anyway.

Reply:
Gauges boundary checked in 3665114

Reply:
Beware, gauge is not a counter, it's just a sampled value of an arbitrary (float/double) number. E.g. CPU temperature:

```
cpu_temp: 78.2|g -> 78.2
cpu_temp: 78.1|g -> 78.1
cpu_temp: 65.0|g -> 65.0
```

No aggregation is done, this is the most simple possible metric.

```